tests/unit/test_hnsw_tiered.cpp

```diff
@@ -2799,3 +2803,261 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
+    }
+    for (size_t i = 0; i < n_labels; i++) {
+        VecSimIndex_DeleteVector(tiered_index, n_labels + i);
+    }
+    // At this point, repair jobs should be executed in the background.
+    EXPECT_EQ(tiered_index->getHNSWIndex()->getNumMarkedDeleted(), n);
+
+    // Insert INPLACE another n vector (instead of the ones that were deleted).
+    VecSim_SetWriteMode(VecSim_WriteInPlace);
+    // Run twice, at first run we insert non-existing labels, in the second run we overwrite them
+    // (for single-value index only).
+    for (auto overwrite : {0, 1}) {
+        for (size_t i = 0; i < n; i++) {
+            TEST_DATA_T vector[dim];
+            for (size_t j = 0; j < dim; j++) {
+                vector[j] = std::rand() / (TEST_DATA_T)RAND_MAX;
+            }
+            EXPECT_EQ(tiered_index->addVector(vector, i % n_labels + n_labels),
+                      TypeParam::isMulti() ? 1 : 1 - overwrite);
+            // Run a query and see that we only receive ids with label < n_labels+i
+            // (the label that we just inserted), and the first result should be this vector.
+            auto ver_res = [&](size_t label, double score, size_t index) {
+                if (index == 0) {
+                    EXPECT_EQ(label, i % n_labels + n_labels);
+                    EXPECT_DOUBLE_EQ(score, 0);
+                }
+                if (!overwrite) {
+                    ASSERT_LE(label, i + n_labels);
+                }
+            };
+            runTopKSearchTest(tiered_index, vector, 10, ver_res);
+        }
+    }
+
+    thread_pool_join(jobQ, run_thread);
+    ASSERT_EQ(tiered_index->frontendIndex->indexSize(), 0);
+    ASSERT_EQ(tiered_index->backendIndex->indexLabelCount(), 2 * n_labels);
+
+    delete index_ctx;
+}
+
+TYPED_TEST(HNSWTieredIndexTest, bufferLimit) {
+    // Create TieredHNSW index instance with a mock queue.
+    size_t dim = 4;
+    HNSWParams params = {.type = TypeParam::get_index_type(),
+                         .dim = dim,
+                         .metric = VecSimMetric_L2,
+                         .multi = TypeParam::isMulti()};
+    VecSimParams hnsw_params = CreateParams(params);
+    auto jobQ = JobQueue();
+    auto index_ctx = new IndexExtCtx();
+    size_t memory_ctx = 0;
+
+    // Create tiered index with buffer limit set to 0.
+    auto *tiered_index = this->CreateTieredHNSWIndex(hnsw_params, &jobQ, index_ctx, &memory_ctx,
+                                                     DEFAULT_PENDING_SWAP_JOBS_THRESHOLD, 0);
+    auto allocator = tiered_index->getAllocator();
+
+    GenerateAndAddVector<TEST_DATA_T>(tiered_index, dim, 0);
+    ASSERT_EQ(tiered_index->backendIndex->indexSize(), 1);
+    ASSERT_EQ(tiered_index->frontendIndex->indexSize(), 0);
+    ASSERT_EQ(tiered_index->labelToInsertJobs.size(), 0);
+
+    // Set the flat limit to 1 and insert another vector - expect it to go to the flat buffer.
+    tiered_index->flatBufferLimit = 1;
```

Comment:
don't we need it in the parameters?

Reply:
We have it. Only in this test, I set the field directly.